Move unboxed dispatch decision into dispatcher

smessmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #28494 [wip] Implement backend level fallback for c10
* #28314 Simplify c10 dispatcher
* **#28251 Move unboxed dispatch decision into dispatcher**

Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

smessmer committed:
Move unboxed dispatch decision into dispatcher

Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/core/dispatch/DispatchTable.h:
```diff
@@ -139,20 +139,33 @@ class DispatchTable final {
+   template<class... Args>
+   const KernelFunction& lookup(const Args&... args) const {
+     return lookup_([&] () -> c10::optional<TensorTypeId> {
+       if (!dispatch_strategy_.is_valid_) {
```
Still not sure why we need to indirect through a dispatch strategy

smessmer replied:
I'm currently working on another simplification diff that will probably address that

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

dzhulgakov reviewed (approved):
nice and clean!

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 4e5398d:
* 0/3 flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=4e5398d4d5e1969c4ac7db44f1f8375e76b36f29).

---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

smessmer committed:
Update on "Move unboxed dispatch decision into dispatcher"


Before, the dispatch key for unboxed operators from native_functions.yaml was generated in codegen and passed to the c10 dispatcher.
Now, we generate it inside of the dispatcher, right next to where the same thing happens for boxed calls.

Differential Revision: [D17989951](https://our.internmc.facebook.com/intern/diff/D17989951/)

[ghstack-poisoned]

ezyang commented:
This diff appears to reliably cause Windows to run out of memory when linking:

```
19:18:16    Creating library LIB\torch.lib and object LIB\torch.exp
19:18:16 LINK : fatal error LNK1102: out of memory
19:18:16 ninja: build stopped: subcommand failed.
19:18:16 -- Building version 1.4.0a0+88a34ef
```

I verified the last three CI results on the PR all failed this way. I am reverting it to give us more headroom on master.
